### PR TITLE
[DataLoader] Make distributed lazily initialized & share seed via PG

### LINKED ISSUE #84492: The distributed setting call in data loader constructor is invalid
### 🐛 Describe the bug

In the following data loader constructor for IterDataPipe (torchdata) dateset, it calls the `_get_distributed_settings` to get the world_size and rank. While in general, the process group is not yet initialized when the data loader constructor is called in the main process. So it will just return ws=1 and rank=0.

Should we put the dist call within the `_sharding_worker_init_fn` when the process group is already initialized? 

```
        if isinstance(self.dataset, IterDataPipe):
            self.dataset = _IterDataPipeSerializationWrapper(self.dataset)
            ws, rank = _get_distributed_settings()
            if num_workers > 0:
                self.worker_init_fn = functools.partial(
                    _sharding_worker_init_fn, self.worker_init_fn, ws, rank)
            else:
                torch.utils.data.graph_settings.apply_sharding(self.dataset, ws, rank)
```

### Versions

```
>>> import torch
>>> torch.__version__
'1.12.0+cu113'
```
cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @SsnL @VitalyFedyunin @ejguan @NivekT @asaluja

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #84492 https://github.com/pytorch/data/issues/772

## Changes
- Move the logic of distributed sharding from the constructor of DataLoader to the constructor of DataLoaderIterator. This would prevent the Error caused by lazy distributed process initialization
- Replace distributed store by process group (`gloo`) to share the random seed because `mpi` backend doesn't provide distributed store.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -631,6 +575,9 @@ def __init__(self, loader: DataLoader) -> None:
         self._drop_last = loader.drop_last
         self._index_sampler = loader._index_sampler
         self._num_workers = loader.num_workers
+        ws, rank = _get_distributed_settings()
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:578]: Change 1: Move distributed information to the constructor of `DataLoaderIterator`.

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -565,63 +556,16 @@ def _create_warning_msg(num_worker_suggest, num_worker_created, cpuset_checked):
...
 
 class _BaseDataLoaderIter(object):
     def __init__(self, loader: DataLoader) -> None:
         self._dataset = loader.dataset
-        self._shared_seed = loader._get_shared_seed()
+        self._shared_seed = None
+        self._pg = None
         if isinstance(self._dataset, IterDataPipe):
+            if dist.is_available() and dist.is_initialized():
+                self._pg = dist.new_group(backend="gloo")
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:567]: Change 2: Use process group rather than store

File: torch/utils/data/dataloader.py [hunk 3]
```diff
@@ -119,6 +117,13 @@ def _sharding_worker_init_fn(worker_init_fn, world_size, rank_id, worker_id):
         worker_init_fn(worker_id)
 
 
+def _share_dist_seed(generator, pg):
```
Comment (NivekT) [inline torch/utils/data/dataloader.py:120]: Question: so this handles everything that `_get_shared_seed` did? Including `seed_counter`, `timeout`, `check_interval`?

**Review by NivekT — APPROVED** [review]
LGTM! Just some questions:

I forgot where the existing tests are for these, can you point them out?

File: torch/utils/data/dataloader.py (hunk 2 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:567]: It doesn't as `gloo` is required whenever PyTorch distributed is compiled.

File: torch/utils/data/dataloader.py (hunk 3 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:120]: Yep. ProcessGroup itself contains `timeout` and `check_interval`. And, the counter would rely on `broadcast` function.

**Comment by ejguan** [comment]
This PR has been validated by the newly introduced tests in https://github.com/pytorch/data/pull/727

**Review by VitalyFedyunin — APPROVED** [review]
By any chance, we can include a test case for it? (I especially concerned about testing same thing for DLV2)

**Comment by ejguan** [comment]
> By any chance, we can include a test case for it? (I especially concerned about testing same thing for DLV2)

I was thinking about it. But, I add the tests for both DLv2 and DL in https://github.com/pytorch/data/pull/727 because I find the most of code are duplicated. WDYT? Should I move it to PyTorch Core?

**Comment by VitalyFedyunin** [comment]
No need. Missed the fact that https://github.com/pytorch/data/pull/727 is also testing lazy init case.
